reps/2023-07-20-torch-trainer-apis.md

```diff
@@ -0,0 +1,371 @@
+    # Restore checkpoint.
+    checkpoint_path = None
+    checkpoint = ray.train.get_context().get_checkpoint()
+    if checkpoint:
+        checkpoint_dir = checkpoint.to_directory()
+        checkpoint_path = Path(checkpoint_dir) / CHECKPOINT_FILE_NAME
+    trainer.train(resume_from_checkpoint=checkpoint_path)
+```
+
+### HuggingFace Accelerate
+
+At its core, HuggingFace Accelerate can be used by configuring and instantiating an `Accelerator` object in your training code.
+
+HuggingFace Accelerate also provides two **optional** CLI commands:
+1. `accelerate config` - Generates a configuration file that is used to configure the `Accelerator` in addition to distributed launching parameters.
+2. `accelerate launch` - Launches a distributed training job using the configuration file generated by `accelerate config`.
+
+The current `AccelerateTrainer` provides functionality similar to `accelerate launch`, in which it will read the generated configuration file and apply it to the `Accelerator`. However, as Ray Train already provides a distributed launching mechanism with its own configurability, we find diminishing value in parsing the configuration file simply for configuring the `Accelerator`. Additionally, it has been a common misconception that the user _must_ use `AccelerateTrainer` in order to use Accelerate. As such, in this proposal we simplify the story by recommending that users configure the `Accelerator` directly in their `TorchTrainer` training code.
+
+#### Accelerate
+```python
+from accelerate import Accelerator
+
+accelerator = Accelerator(**accelerator_kwargs)
+...
+```
+
+#### `AccelerateTrainer` (Current)
+
+```bash
+! accelerate config
+```
+
+```python
+from accelerate import Accelerator
+from ray.train.accelerate import AccelerateTrainer
+
+def train_func(config):
+    accelerator = Accelerator()
+    ...
+
+trainer = AccelerateTrainer(
+    train_loop_per_worker=train_func,
+    accelerate_config=accelerate_config,
+    ...
+)
+trainer.fit()
+```
+
+#### `TorchTrainer` (Proposed)
+
+In this proposal, we recommend configuring the `Accelerator` object directly in the training code, exactly as you would if you were not using Ray Train.
+
+```python
+from accelerate import Accelerator
+from ray.train.torch import TorchTrainer
+
+def train_func(config):
+    accelerator = Accelerator(**accelerator_kwargs)
+    ...
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_func,
+    ...
+)
+trainer.fit()
+```
+
+
+**Reporting & Checkpointing:**
+
+This is done in the same way as the aforementioned Torch example.
+
+## Compatibility, Deprecation, and Migration Plan
+
+Ray 2.7: 
+- Add and expose the public utilities for making training framework code compatible with Ray Train.
+- Update documentation and examples to use the `TorchTrainer` API.
+- Mark `AccelerateTrainer`, `TransformersTrainer`, and `LightningTrainer` APIs as deprecated.
+
+Ray 2.8:
+- Remove `AccelerateTrainer`, `TransformersTrainer`, and `LightningTrainer` APIs.
```

Comment:
Nit, per convention we should print a warning in 2.7, raise an error in 2.8, and remove the API in 2.9.